### PR TITLE
(329) HDC & CRD date validations

### PR DESCRIPTION
**Pull request by libuk**
[JIRA](https://dsdmoj.atlassian.net/browse/CAS2-329?atlOrigin=eyJpIjoiMDhiYThhOWQ3NDliNDJlZmE2ODM0Yjk4NTNlZjAzMjYiLCJwIjoiaiJ9)

# Context
As a referrer
I want to be stopped from making entry errors when inputting HDC and CRD dates
So that I do not get chased for corrections

# Changes in this PR
* Introduce the following validations for the HDC page
  * HDC eligibility date cannot be more than 6 months before the conditional release date (CRD)
  * HDC eligibility date must be before the conditional release date (CRD)
  * Conditional release date (CRD) must be the same as or after today’s date

## Screenshots of UI changes
Screenshot 2024-03-19 at 16 19 48
Screenshot 2024-03-19 at 16 19 22
Screenshot 2024-03-19 at 16 19 06

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Are any changes required to the e2e tests?
- [ ] If you've added a new route, have you added a new
  `auditEvent`? (see `server/routes/apply.ts` for examples)
- [ ] Are there environment variables or other infrastructure configuration which needs to be included in this release?
- [ ] Are there any data migrations required. Automatic or manual?
- [ ] Does this rely on changes being deployed to the CAS API?

## Post merge checklist

Once we've merged it will be auto-deployed to the dev environment.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-community-accommodation-tier-2-ui).

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Review by Harriethw — APPROVED** [review]
Thanks for making such clear commits and for doing so much date wrangling. (and 100% unit test coverage 🥳 )

**Review by Gweaton — APPROVED** [review]
Nice one, I like the use of JS docs on this project, haven't used them much up til now!

File: server/utils/dateUtils.ts [hunk 1]
```diff
@@ -95,18 +95,22 @@ export class DateFormats {
...
    * @param dateInputObj an object with date parts (i.e. `-month` `-day` `-year`), which come from a `govukDateInput`.
    * @param key the key that prefixes each item in the dateInputObj, also the name of the property which the date object will be returned in the return value.
    * @returns a friendly date.
    */
-    static dateAndTimeInputsToUiDate(dateInputObj: Record<string, string>, key: string | number, format = 'medium' as 'short' | 'long' | 'medium') {
-      const iso8601Date = DateFormats.dateAndTimeInputsToIsoString(dateInputObj, key)[key]
-  
-      return DateFormats.isoDateToUIDate(iso8601Date, { format: format })
-    }
+  static dateAndTimeInputsToUiDate(
```
[the remark this answers is not in the document]
Comment (libuk) [reply server/utils/dateUtils.ts:116]: Yes sorry, this formatting change must've sneaked in

[the remark this answers is not in the document]
Comment (libuk) [reply server/utils/dateUtils.ts:116]: > I like the use of JS docs on this project, haven't used them much up til now!

Neither have I, but I find them really helpful for utils.

**Pull request merged**